Make projected parquet collection schema forward compatible with the given file schema

mickjermsurawong-stripe opened the pull request:
- Problem: projection schema is created by [ThriftSchemaConvertVisitor](https://github.com/apache/parquet-mr/blob/master/parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftSchemaConvertVisitor.java#L180) which suffix fields with `_tuple` 
https://github.com/twitter/scalding/blob/59d932731f7396eaaf6024624d1ce6660534ca77/scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ScroogeReadSupport.java#L98-L101
- When reading parquet file with standard format 3-level or even legacy ones 2-level `array`, projection fails because it has name `_tuple` which is incompatible. For example, given file schemas to represent `List<String> country_codes`
```
message compliant_schema {
  required group country_codes (LIST) {
    repeated binary array (UTF8);
  }
  required int32 to_be_removed;
}
```
will fail with the generated projection 
```
message thrift_generated {
  optional group country_codes (LIST) {
    repeated binary country_codes_tuple (UTF8);
  }
}
```
- This PR introduces `ParquetListFormatForwardCompatibility` to convert thrift-generated schema to compliant ones. The difficulties here are:
  - we can not simply just traverse the two graphs and swap node one-for-one. The node has to be from the projection, because it has subset of fields and different optional/required info.
  - there are at least 4 variations of compliant parquet types, and we want to convert from non-compliant to compliant, and among the compliant. 

- The strategy here is to implement different rules for each format. The rule tell us how to decompose the list schema, and to reconstruct it again.
  - For example, we match projected schema with Rule 1, and match file schema with Rule 2.
  - Rule 1 will decompose the projected schema, and Rule 2 will take that information to construct in light of the file schema.

mickjermsurawong-stripe committed:
forward list compat

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/twitter/scalding?pullRequest=1921) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/twitter/scalding?pullRequest=1921) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mickjermsurawong-stripe<br/>:x: joshrosen-stripe<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/twitter/scalding?pullRequest=1921) it.</sub>

mickjermsurawong-stripe commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+   * @param projection projection type which has legacy format
+   * @return projection schema in the new format.
+   */
+  public Type resolveTypeFormat(Type fileType, Type projection) {
```
Main logic of this PR

johnynek replied:
can't this be a static method?

mickjermsurawong-stripe commented on scalding-parquet-scrooge/src/test/scala/com/twitter/scalding/parquet/scrooge/ScroogeReadSupportTests.scala:
```diff
@@ -90,6 +90,384 @@ class ScroogeReadSupportTests extends WordSpec with Matchers with HadoopSharedPl
     }
   }
 
+  "ScroogeReadSupport resolving list format" should {
```
Note that in projected schema, the field is optional, and that is maintained in the result.

mickjermsurawong-stripe commented on scalding-parquet-scrooge/src/test/scala/com/twitter/scalding/parquet/scrooge/ScroogeReadSupportTests.scala:
```diff
@@ -90,6 +90,384 @@ class ScroogeReadSupportTests extends WordSpec with Matchers with HadoopSharedPl
+          |message SampleProjection {
+          |  optional group foo (LIST) {
+          |    repeated group foo_tuple (LIST) {
+          |      repeated binary foo_tuple_tuple (UTF8);
```
Got this schema from `ThriftSchemaConverter` which is what `ScroogeReadSupport` does when projection given thrift class.
```
      val innerListType = new ListType(new ThriftField("list", 2, Requirement.REQUIRED, new ThriftType.StringType))
      val listType = new ListType(new ThriftField("list", 2, Requirement.REQUIRED, innerListType))
      val children = new ThriftField("foo", 1, Requirement.REQUIRED, listType)
      val message = new ThriftSchemaConverter().convert(new StructType(util.Arrays.asList(children), StructOrUnionType.STRUCT))
```

mickjermsurawong-stripe commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+   * Rule describes how to match a repeated type, how to decompose them, and reconstruct a
+   * repeated type.
+   */
+  abstract static public class Rule {
```
This and bulk of below are implementation of different rules.

mickjermsurawong-stripe commented:
cc @moulimukherjee

mickjermsurawong-stripe commented:
Hi @johnynek,
We ran into a problem where Scalding job fails reading parequet files produced by another Spark job. The problem we discovered is due to projected schema which is generated from Thrift class has different list format than our Spark job.

Just would like to ask for a quick look whether this is the right strategy, and if this is something that we would like to push it to the public one as well.

johnynek commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+ * {{@code org.apache.parquet.thrift.ThriftSchemaConvertVisitor}} which always suffix
+ * list element with "_tuple".
+ */
+public class ParquetListFormatForwardCompatibility {
```
we generally prefer scala in this repo. Was there any reason to implement this in Java?

johnynek commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+    @Override
+    public Type createCompliantRepeatedType(Type type, String name, Boolean isElementRequired, OriginalType originalType) {
+      if (!isElementRequired) {
+        throw new IllegalArgumentException("Rule 1 can only take required element");
```
you use the string "Rule 1" here, but the class name is `RulePrimitiveElement`. Can you explain the difference?

mickjermsurawong-stripe replied:
Sorry, that's bad naming. Will clean up.

johnynek commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+    return matchedRule;
+  }
+
+  private Type makeForwardCompatible(Type repeatedFileType, Type repeatedProjectedType) {
```
can't this be a static method?

johnynek commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+    }
+  }
+
+  private Rule findFirstRule(Type repeatedType, String debuggingTypeSource) {
```
can't this be a static method?

johnynek commented on scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ParquetListFormatForwardCompatibility.java:
```diff
@@ -0,0 +1,349 @@
+    if (projection.isPrimitive() || fileType.isPrimitive()) {
+      return projection;
+    }
+    ParquetListFormatForwardCompatibility compatibility = new ParquetListFormatForwardCompatibility();
```
why do we need to allocate a new instance?

johnynek reviewed (commented):
Thanks for the PR.

Have we run this internally to do some smoke tests that it can read old data that was previously written?

I added some general comments. I really don't know the latest with parquet. Maybe @isnotinvain or @julienledem can comment?

cc @ttim

johnynek commented:
Also, you need to bump the `.travis.yml` jdk setting to `openjdk8` since `oraclejdk8` I think no longer works on travis.

mickjermsurawong-stripe commented:
Thanks @johnynek. Ah I haven't done any tests internally on actual data. Just purely unit tests now. Will follow up with that. 

I can port all these to Scala, and will make the resolver stateless; yup the object instantiation isn't neccessary.

mickjermsurawong-stripe committed:
bump travis to openjdk8

mickjermsurawong-stripe committed:
fail when projecting on list of struct

mickjermsurawong-stripe committed:
recurse only on elements

mickjermsurawong-stripe committed:
check for non-optional extra fields for projection

mickjermsurawong-stripe committed:
move to scala

mickjermsurawong-stripe committed:
migrate test to different class

mickjermsurawong-stripe committed:
handle map legacy format

mickjermsurawong-stripe committed:
format import

mickjermsurawong-stripe committed:
improve docs

mickjermsurawong-stripe committed:
fix style warning

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+
+object ParquetCollectionFormatForwardCompatibility {
+
+  private val LOGGER = LoggerFactory.getLogger(getClass)
```
Nit: `private[this] val logger = LoggerFactory.getLogger(getClass)` seems to be the convention used elsewhere in the existing code.

mickjermsurawong-stripe replied:
Addressed.

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+   */
+  def formatForwardCompatibleMessage(sourceType: MessageType, targetType: MessageType): MessageType = {
+    val groupResult = formatForwardCompatibleType(sourceType, targetType).asGroupType()
+    LOGGER.info("Making source schema to be compatible with target" +
```
Should this be `info` level by default? Or will that be too verbose? I took a look at the existing code to try to find precedent but couldn't identify an obvious existing convention.

isnotinvain replied:
I think too verbose, the schemas can often be very large. Maybe debug level?

mickjermsurawong-stripe replied:
Updated to debug.

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+                                    repeatedListType: Option[Type] = None,
+                                    repeatedMapType: Option[Type] = None)
+
+  private def unwrapGroup(typ: Type, wrappers: Seq[GroupType] = Seq()): GroupUnwrapped = {
```
Perhaps a naive question, but what does "wrapped" or "wrapper" refer to in this case? Could you add a comment explaining this? That might make the following code easier to review.

mickjermsurawong-stripe replied:
I now simplified to just extract for only the repeated type, and let the recursive call in main method `formatForwardCompatibleType` handle multiple layers instead.
This now becomes a sealed trait for the group type itself, and a helper method to get its repeated type
```
private[scrooge] sealed trait CollectionGroup {
  def groupType: GroupType

  def repeatedType: Type
}
```

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+    }
+  }
+
+  def isGroupList(projection: Type): Boolean = {
```
To confirm my understanding:

This is matching on things like 

```
required group my_list (LIST) {
  repeated group list {
    optional binary element (UTF8);
  }
}
```

where `groupProjection` here refers to the top-most type, and the `groupProjection.getFields.get(0).isRepetition(Type.Repetition.REPEATED)` check would apply to the second-level element (`repeated group list`)?

mickjermsurawong-stripe replied:
yup that's correct on `groupProjection`, and the repeated is the second level.

The part after `repeated` could be different depending on formats eg.
```
required group my_list (LIST) {
  repeated repeated int32 [element|array];
}
```

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+        repeatedListType=None,
+        repeatedMapType=None
+      )
+    } else if (typ.asGroupType.getFieldCount != 1) {
```
In this `else if` branch and the `else` branch I noticed we have `typ.asGroupType` calls everywhere. To avoid that and to make it clear that `!typ.isPrimitive` implies that `typ` is a group type, could we instead do something like

```scala
if (typ.isPrimitive) {
   // ...
} else {
  val groupTyp = typ.asGroupTyp
  if (groupTyp.getFieldCount != 1) {
     //...
  } else {
     //...
  }
}
```

We might be able to further improve clarity by changing the sign of the second condition, doing

```scala
if (typ.isPrimitive) {
   // ...
} else {
  val groupTyp = typ.asGroupTyp
  if (groupTyp.getFieldCount == 1) {  // <--- inverted condition
     //...
  } else {
     //...
  }
}
```

which then simplifies or clarifies the

```
// Note the field count is strictly 1 here
```

comment present in the current code.

We could then add a comment in the `else` branch saying something like

```
// The group contains more than one field, so ... <whatever that implies>
```

This is maybe nitpicky, but I'm trying to walk through each case here to make sure that I understand what all of the branches correspond to, so I'm considering whether a different ordering or presentation of the cases can make it easier to understand and verify.

mickjermsurawong-stripe replied:
Simplified this to two methods to extract group list and maplist instead.

mickjermsurawong-stripe commented:
hi  @isnotinvain or @julienledem, I've addressed the first round of general feedback, and added support for map schema compat as well. Would appreciate your review here.

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+   * this is because the subset fields of source node and its optional/required must
+   * be maintained in the formatted result.
+   */
+  private def formatForwardCompatibleType(sourceType: Type, targetType: Type): Type = {
```
I find this method's pattern match to be a bit difficult to intuitively understand and I think that maybe it's because of the `GroupUnwrapped`: there, we seem to have a big cross product of potential possibilities, but it looks like we only have explicit cases for some of them.

By construction, it looks like we never wind up with `repeatedListType` and `repeatedMapType` both non-None.

To simplify the pattern match and `unwrapGroup`, what do you think about instead defining a `sealed trait` of case-classes corresponding to the different cases that we might need to handle? What that approach, I think we can avoid the need to do this more complicated positional pattern-matching. 

Given that, could it make sense to have methods like `ParquetMapFormatRule.isGroupMap()` instead be "extractor"-type methods which return `Option[GroupMapSomethingIHaventNamedYet]`? That could help to avoid separate definition and extraction code that we have now (where the "is X" and "extract X" code for the same logical concept are in separate places).
 
This would also have the advantage of avoiding the need to have comments like `// Format list` to explain the cases, since that semantic information could be instead carried in the sealed trait class names.

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+        val sourceGroup = sourceType.asGroupType
+        val targetGroup = targetType.asGroupType
+
+        import scala.collection.JavaConverters._
```
Import at the top of the file? That should be safe because this is `Converters` (not `Conversions`, which is much more implicit).

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+    ruleFound
+  }
+
+  def wrapElementAsRepeatedType(rule: ParquetListFormatRule, repeatedType: Type, elementType: Type): Type = rule.createCompliantRepeatedType(
```
This is formatted a bit funny right now; putting the `rule.create*` on the next line?

joshrosen-stripe reviewed (commented):
I left some initial review comments, primarily focused on understanding (and potentially simplifying) two of the larger and more complex pattern matches at the heart of the converter.

I also left a couple of minor style nits, but those can be addressed later (in a final polish pass) once we're confident in the rest of the design.

I haven't finished looking at this yet, though: I still need to go through `ParquetListFormatRule` and also need to take a closer look at the test cases.

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+
+  private[scrooge] def isElementRequired(repeatedType: Type): Boolean
+
+  private[scrooge] def check(typ: Type): Boolean
```
From the name, it's a bit non-obvious what this method does. It this actually something like `appliesToType(typ: Type)` which returns `true` if the type is a list of this format and `false` otherwise?

thomaschow replied:
+1 on naming this something more descriptive

mickjermsurawong-stripe replied:
Updated as suggested

thomaschow commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+  def name: String
+}
+
+private[scrooge] object Source extends SourceOrTarget {
```
what's the advantage of distinguishing between Source and Target? is it so we can define supported directionalities in terms of parquet spec conversion? I'm wondering if we're being too specific here and wind up seeing a direction that we didn't engineer for, and we fail

thomaschow replied:
I guess since scalding is going to be stuck on the older parquet spec, this would not happen.

mickjermsurawong-stripe replied:
~~ack. Will see how I can do end-to-end testing in this repo.~~
Ah the only thing i don't support here is the target type of `_tuple` format which then would be compatible with the thrift-generated projection source

thomaschow commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+  override def check(repeatedType: Type): Boolean = repeatedType.getName.endsWith("_tuple")
+
+  override def elementName(repeatedType: Type): String = {
+    repeatedType.getName.substring(0, repeatedType.getName.length - 6)
```
nit: would like docs or variable/method renaming to explain this substring indexing a little more clearly

mickjermsurawong-stripe replied:
I now made `_tuple` as a variable where length is calculated from it more explicitly.

thomaschow commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatRule.scala:
```diff
@@ -0,0 +1,262 @@
+    } else {
+      new GroupType(
+        repetition,
+        "element",
```
nit: DRY with https://github.com/twitter/scalding/pull/1921/files#diff-f1236da4ace7eaf46e33ad1d12d599b8R241 ?

thomaschow replied:
nvm, doesn't seem necessary

thomaschow commented on scalding-parquet-scrooge/src/test/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibilityTests.scala:
```diff
@@ -0,0 +1,919 @@
+import org.apache.parquet.thrift.struct.ThriftType.{ListType, MapType, StructType}
+import org.scalatest.{Matchers, WordSpec}
+
+class ParquetCollectionFormatForwardCompatibilityTests extends WordSpec with Matchers {
```
ooc how hard would it be to add some actual thrift fixtures based tests in here?

mickjermsurawong-stripe replied:
ack. Will see how I can do end-to-end testing in this repo.

isnotinvain commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,153 @@
+
+import scala.collection.mutable
+
+object ParquetCollectionFormatForwardCompatibility {
```
Parquet is complicated because there are always 4 schemas involved: the write-time schema, the read-time schema, and a projected schema in terms of the write-schema, and a projected schema in terms of the read-time schema. I find this really hard to keep track of, so I think it'd be a huge help to me and future readers to put a really detailed comment on this class explaining what it's for, what it makes compatible, what the problem is, how it solves it, etc. I think for something like this, the more detailed the better.

isnotinvain reviewed (commented):
Thanks for doing this, parquet's interop story between thrift and other object models often doesn't work for reasons like this. It might be worth making sure that this is all that's required to make it work, it could be that after fixing this there will still be more things that don't interop between parquet-thrift and other parquet integrations. 

I will try to dig into the details of this review, but right away I can say that I think all this new code should come with new tests. The tests should also help us review by demonstrating how it ought to work.

mickjermsurawong-stripe commented:
hi @isnotinvain thank you for the note. 
Will go through comments by my colleagues and improve docs as you suggested. 
The test class is in `ParquetCollectionFormatForwardCompatibilityTests.scala` not sure if that's what you are looking for, but that demonstrates the format conversion here.

mickjermsurawong-stripe committed:
address PR feedback

mickjermsurawong-stripe committed:
undo import wildcard

mickjermsurawong-stripe committed:
fix warning descenents of sealed traits

mickjermsurawong-stripe committed:
remove duplicate test

mickjermsurawong-stripe committed:
improve test names and remove one duplicate

mickjermsurawong-stripe committed:
add docs

mickjermsurawong-stripe commented:
@joshrosen-stripe, @tchow-stripe 
Thanks for the really helpful review. I've simplified many parts of the code, and addressed the PR feedback.
Pending issues are figuring out to do end-to-end testing here.
@isnotinvain this is ready for another round of review.

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,130 @@
+    sourceGroup.groupType.withNewFields(formattedRepeated)
+  }
+
+  private def findCollectionGroup(typ: Type): Option[CollectionGroup] = {
```
Naming nit:

```suggestion
  private def extractCollectionGroup(typ: Type): Option[CollectionGroup] = {
```

"find" implies that the return values will be a subset of the input, while "extract" (which the underlying functions use) suggests that the whole input will be considered.

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,130 @@
+        formatForwardCompatibleCollectionGroup[ListGroup](sourceGroup, targetGroup)
+      case (Some(sourceGroup: MapGroup), Some(targetGroup: MapGroup)) =>
+        formatForwardCompatibleCollectionGroup[MapGroup](sourceGroup, targetGroup)
+      case _ if sourceType.isPrimitive || targetType.isPrimitive => // Base case
```
I would explicitly check and throw if one is a primitive but not the other. As written this will silently pass through certain schema mismatches.

mickjermsurawong-stripe replied:
Ack

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,130 @@
+   * this is because the subset fields of source node and its optional/required must
+   * be maintained in the formatted result.
+   */
+  private def formatForwardCompatibleType(sourceType: Type, targetType: Type): Type = {
```
```suggestion
  private def formatForwardCompatibleType(fileType: Type, projectionType: Type): Type = {
```

"source" and "target" are very generic names and this function is really only used for one thing. It might make sense to rename them accordingly to make reasoning about the correctness here easier.

mickjermsurawong-stripe replied:
From conversation thread below: 
> By that i mean we want to convert projection source schema to be the in the format of the target file schema.

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,130 @@
+        val targetGroup = targetType.asGroupType
+        val resultFields = sourceGroup.getFields.asScala.map { sourceField =>
+          if (!targetGroup.containsField(sourceField.getName)) {
+            if (!sourceField.isRepetition(Repetition.OPTIONAL)) {
```
This check is incorrect, I think. "source" == the file's schema, yes (see above renaming suggestion)? If so, then the optionality of a field which is not included in the projection is not important. We should be iterating over the fields in the projected (target) schema and checking the optionality of the projected fields if they're missing from the file.

mickjermsurawong-stripe replied:
Ah no, source here is actually the projection, and target is the file's scheme.
By that i mean we want to convert projection source schema to be the in the format of the target file schema. 
Given this confusion, i think makes sense that I should rename this.

isnotinvain replied:
on this point, I think naming convention along the lines of: fileSchema, readSchema, projectedFileSchema, projectedReadSchema all help clarify which is which

xton reviewed (commented):
This review is incomplete, I'll finish later this evening. I've got some rename suggestions and some possible bugs so far.

mickjermsurawong-stripe committed:
Check on schema type mismatch

mickjermsurawong-stripe committed:
explicit rename from source/target to projected read schema and file schema

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,146 @@
+   * @return formatted result
+   */
+  def formatForwardCompatibleRepeatedType(sourceRepeatedType: Type,
+                                          targetRepeatedType: Type,
```
Another rename candidate?

mickjermsurawong-stripe replied:
Changed to `fileRepeatedType` and `readRepeatedType`

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatter.scala:
```diff
@@ -0,0 +1,270 @@
+  private val LOGGER = LoggerFactory.getLogger(getClass)
+
+  def formatForwardCompatibleRepeatedType(repeatedSourceType: Type,
+                                          repeatedTargetType: Type,
```
another rename candidate?

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,146 @@
+      fileGroup.repeatedType,
+      projectFileType(_, _))
+    // Respect optional/required from the projected read group.
+    projectedReadGroup.groupType.withNewFields(projectedFileRepeatedType)
```
Could this end up letting through cases where the file's field is optional and the projection's field is required? If so we should explicitly detect and throw for that (probably in `projectFileType`).

mickjermsurawong-stripe replied:
Added the assertion at the "struct projection" which will have access of field-level optional/required

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,146 @@
+        projectFileGroup[ListGroup](projectedReadGroup, fileGroup)
+      case (Some(projectedReadGroup: MapGroup), Some(fileGroup: MapGroup)) =>
+        projectFileGroup[MapGroup](projectedReadGroup, fileGroup)
+      case _ => // Field projection
```
Comment nit: I might call this "struct projection". We're projecting through many fields.

mickjermsurawong-stripe committed:
support creating _tuple format and generalize compat in all directions

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetListFormatter.scala:
```diff
@@ -0,0 +1,270 @@
+  override val rules: Seq[ParquetListFormatRule] = Seq(
+    PrimitiveElementRule, PrimitiveArrayRule,
+    GroupElementRule, GroupArrayRule,
+    TupleRule, StandardRule
```
Soooo many cases... That said, I like this approach. Makes it painfully clear the variants that we're supporting.

mickjermsurawong-stripe committed:
support legacy spark list of nullable elements

mickjermsurawong-stripe committed:
improve docs

mickjermsurawong-stripe committed:
add tests for all supported list compat conversions

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetMapFormatter.scala:
```diff
@@ -0,0 +1,65 @@
+   * @param targetRepeatedMapType
+   */
+  def formatForwardCompatibleRepeatedType(sourceRepeatedMapType: Type,
+                                          targetRepeatedMapType: Type,
```
Candidate for rename?

xton commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatForwardCompatibility.scala:
```diff
@@ -0,0 +1,159 @@
+        (ParquetMapFormatter, fieldContext.copy(nestedListLevel = fieldContext.nestedListLevel + 1))
+      case c if c == classOf[ListGroup] =>
+        (ParquetListFormatter, fieldContext.copy(nestedListLevel = fieldContext.nestedListLevel + 1))
+    }
```
Let's refactor this to avoid reflection and implicit magic. E.g.:

```suggestion
private def projectFileGroup(projectedReadGroup: T, fileGroup: T, fieldContext: FieldContext, formatter: ParquetCollectionFormatter): GroupType = {
```

mickjermsurawong-stripe replied:
Updated to
```
def projectFileGroup(fileGroup: CollectionGroup,
                               projectedReadGroup: CollectionGroup,
                               fieldContext: FieldContext,
                               formatter: ParquetCollectionFormatter)
```

xton reviewed (changes requested):
One potential simplification and some rename/docs nits. Looking good!

mickjermsurawong-stripe committed:
improve docs

mickjermsurawong-stripe committed:
remove classtag inference

mickjermsurawong-stripe committed:
rename schema name and use consistent method

mickjermsurawong-stripe committed:
file rename to drop "forward" compat

mickjermsurawong-stripe committed:
test rename and make variables consistent

mickjermsurawong-stripe committed:
make names file/read oriented

mickjermsurawong-stripe committed:
improve test make sure formatted type is still compatible with given file schema

mickjermsurawong-stripe committed:
check for field optional/required

mickjermsurawong-stripe commented:
Thanks @xton, addressed the feedback.
@isnotinvain could you help to take a look on this please?

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatCompatibility.scala:
```diff
@@ -0,0 +1,163 @@
+  private def projectFileType(fileType: Type, projectedReadType: Type, fieldContext: FieldContext): Type = {
+    (extractCollectionGroup(projectedReadType), extractCollectionGroup(fileType)) match {
+      case _ if projectedReadType.isPrimitive && fileType.isPrimitive =>
+        projectedReadType
```
Do we need to check that both types are the _same_ primitive type? Or is that handled elsewhere? I imagine that certain types of primitive-type-changing projections might be supported (e.g. widening an int to a long) but others might not be (e.g. converting a string to an int).

mickjermsurawong-stripe replied:
That type compatibility is handled in a separate place here, so by the time this is called we should already have type compatibility.
https://github.com/twitter/scalding/blob/73ce124f3c0156bd41f5394d2b609a5cd74f370e/scalding-parquet-scrooge/src/main/java/com/twitter/scalding/parquet/scrooge/ScroogeReadSupport.java#L133-L136

joshrosen-stripe commented on scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatCompatibility.scala:
```diff
@@ -0,0 +1,163 @@
+  def formatCompatibleRepeatedType(fileRepeatedType: Type,
+                                   readRepeatedType: Type,
+                                   fieldContext: FieldContext,
+                                   recursiveSolver: (Type, Type, FieldContext) => Type): Type
```
In practice, I think this is always invoked with the `projectFileType` method defined above?

mickjermsurawong-stripe replied:
Yup it is. My thought is to make this recursive call explicit from the caller side so it's clear from the reading within this format compatbility layer instead of jumping into the respective formatter which then cyclically refers back to the caller's method.

joshrosen-stripe committed:
Review suggestions for https://github.com/twitter/scalding/pull/1921 (#1)

* Add more comments; add explicit else conditions to some if cases; lift base case out of pattern match

* Formatting changes

* More minor formatting.

mickjermsurawong-stripe committed:
improve code coverage and remove dead code after restructuring

mickjermsurawong-stripe committed:
auto-format from running sbt test

mickjermsurawong-stripe commented:
hi @isnotinvain, would like to follow-up on this if you have more feedback here please.

mickjermsurawong-stripe commented:
Hi  @isnotinvain, another bump on this please.

Updates from our side: we actually have implemented this in our internally (via subclassing of `ParquetScroogeScheme`), and it does cover production use cases of read Spark job schema.
- list having 3 level, and 2-level in legacy mode, compared to `_tuple` suffixed via generated schema from thrift class.
- map having repeated group named `key_value` instead of `map` 
This will also cover more conversion types as outlined in `ParquetMapFormatter` and `ParquetListFormatter`

Without this patch, scalding job reading Spark job output can either fail hard or falsely read empty collection.
